base/src/Data/Macaw/AbsDomain/JumpBounds.hs

```diff
@@ -661,6 +661,45 @@ addRangePred cns v p =
         _ ->
           Right $ branchAssignRangePred n p
 
+-- | @assertEqPred cns x w c isTrue@ asserts equality @x = BVValue c w@ is either true, or false,
+-- based on `isTrue`.
+assertEqPred ::
+  ( OrdF (ArchReg arch)
+  , HasRepr (ArchReg arch) TypeRepr
+  , MemWidth (ArchAddrWidth arch)
+  , ShowF (ArchReg arch)
+  ) =>
+  -- | Constraints that let us evaluate `x`
+  BlockIntraStackConstraints arch ids ->
+  -- | Value to be constrained
+  Value arch ids (BVType w) ->
+  -- | Bitsize of bitvector c
+  NatRepr w ->
+  -- | Numeric constant x should be equal/different to
+  Integer ->
+  -- | `True` if we should assert equality, `False` if we should assert non-equality
+  Bool ->
+  Either String (BranchConstraints arch ids)
+assertEqPred cns x w c isTrue
+  | isTrue =
+    addRangePred cns (intraStackValueExpr cns x)
+                     (mkRangeBound w (fromInteger c) (fromInteger c))
+  | otherwise = do
+    -- !(x == c) becomes:
+    --                   (x >= c + 1)   when c == 0
+    --   (x <= c - 1) || (x >= c + 1)   otherwise
+    -- because ranges use natural numbers and are inclusive on both ends.
```

Comment:
Do we need a special case for when `c == max`? There is special treatment of this sort in the `EvalApp (BVUnsignedLt (BVValue w c) y)` case in `assertPred`, and I'm not sure if there needs to be similar special treatment here as well.

Reply:
Yes. I'm going to refactor this even more because I still don't like how it looks.

Reply:
I pushed a solution with a more aggressive refactoring, and split in two commits.

Reply:
Looks like I might have made a mistake. Will investigate...

Reply:
Found the bug, I needed to be more careful here:
https://github.com/GaloisInc/macaw/blob/45dd9545e5262d201734a9d77df9baec22aaf902/base/src/Data/Macaw/AbsDomain/JumpBounds.hs#L730-L735